src/main/java/io/cos/cas/osf/web/flow/login/OsfPrincipalFromNonInteractiveCredentialsAction.java

```diff
@@ -627,35 +648,79 @@ private OsfApiInstitutionAuthenticationResult notifyOsfApiOfInstnAuthnSuccess(
             throw new InstitutionSsoFailedException("OSF CAS failed to build JWT / JWE payload for OSF API");
         }
         // Send the POST request to OSF API to verify an existing institution user or to create a new one
+        int statusCode;
+        HttpResponse httpResponse;
         try {
-            final HttpResponse httpResponse = Request.Post(osfApiProperties.getInstnAuthnEndpoint())
+            httpResponse = Request.Post(osfApiProperties.getInstnAuthnEndpoint())
                     .addHeader(new BasicHeader("Content-Type", "text/plain"))
                     .bodyString(jweString, ContentType.APPLICATION_JSON)
                     .execute()
                     .returnResponse();
-            final int statusCode = httpResponse.getStatusLine().getStatusCode();
+            statusCode = httpResponse.getStatusLine().getStatusCode();
             LOGGER.info(
-                    "[OSF API] Notify Remote Principal Authenticated Response: username={} statusCode={}",
+                    "[OSF API] Notify Remote Principal Authenticated Response: username={}, statusCode={}",
                     username,
                     statusCode
             );
-            if (statusCode != HttpStatus.SC_NO_CONTENT) {
-                final String responseString = new BasicResponseHandler().handleResponse(httpResponse);
-                LOGGER.error(
-                        "[OSF API] Notify Remote Principal Authenticated Failed: statusCode={}, body={}",
-                        statusCode,
-                        responseString
-                );
-                throw new InstitutionSsoFailedException("OSF API failed to process CAS request");
-            }
-            return new OsfApiInstitutionAuthenticationResult(username, institutionId);
         } catch (final IOException e) {
-            LOGGER.error(
-                    "[OSF API] Notify Remote Principal Authenticated Failed: Communication Error - {}",
-                    e.getMessage()
-            );
+            LOGGER.error("[OSF API] Notify Remote Principal Authenticated Failed: Communication Error - {}", e.getMessage());
             throw new InstitutionSsoFailedException("Communication Error between OSF CAS and OSF API");
         }
+        // CAS expects OSF API to return HTTP 204 OK with no content if authentication succeeds
+        if (statusCode == HttpStatus.SC_NO_CONTENT) {
+            LOGGER.info("[OSF API] Notify Remote Principal Authenticated Passed: institution={}, username={}", institutionId, username);
+            return new OsfApiInstitutionAuthenticationResult(username, institutionId);
+        }
```

Comment:
Flattened the control flow and brought the API success return statement to the top.